Google Sheets V2 documentation (#1038)
rudderlabs/rudderstack-docs: docs/sources/extract/google-sheets-v2.mdx

---
title: "Google Sheets V2"
description: Step-by-step guide on ingesting data from Google Sheets into RudderStack.
aliases: ["Google Sheets source"]
---

> [Google Sheets](https://www.google.com/sheets/about/) is a popular spreadsheet program that lets you create spreadsheets which are updated and saved automatically.

Author: ```suggestion
[Google Sheets](https://www.google.com/sheets/about/) is a popular spreadsheet program that lets you create and manage your spreadsheets.
```

</div>

<GhBadge
  label={'Stability'}
  message={'Beta'}
  color={'blueviolet'}
/>

## Getting started

To set up Google Sheets as a source in RudderStack, follow these steps:

1. Log into your [RudderStack dashboard](https://app.rudderstack.com/).
2. Go to **Sources** > **New source** > **Cloud Extract** and select **Google Sheets V2** from the list of sources.
3. Assign a name to your source and click **Next**.

### Connection settings

Next, configure the following dashboard settings:

<img src="../../assets/cloud-extract-sources/googlesheets-beta-connection-settings.png" alt="Configuring Google Sheets" />

- **Authentication**: From the dropdown, select the authentication mechanism for RudderStack to connect to the Google Sheets API.
    - **Authenticate via Google (OAuth)**: To authenticate via OAuth, click the **Sign in with Google** button, select your Google account, and give RudderStack the required permissions.
    - **Service Account Key Authentication**: Enter your <a href="https://cloud.google.com/iam/docs/creating-managing-service-account-keys#creating">service account</a> JSON credentials in the **Service Account Information** field.

> - **Spreadsheet Link**: From the dropdown, select the spreadsheet from which RudderStack ingests and syncs the data.

Reviewer: ```suggestion
- **Spreadsheet Link**: From the dropdown, select the spreadsheet from which RudderStack should ingest and sync the data.
```

<div class="infoBlock">
<ul>
  <li>For the <strong>Authenticate via Google (OAuth)</strong> authentication option, RudderStack will automatically populate all the spreadsheets associated with the account.</li>
  <li>For the <strong>Service Account Key Authentication</strong> option, you need to first provide access to the required spreadsheet. <strong>Only the spreadsheets for which you provide the access will be listed in the dropdown</strong>. For more information, refer to the <Link to="#faq">FAQ</Link> section below.</li>
</ul>
</div>

> 

Reviewer: **Row batch size** setting and its description can be added..

Author: Actually this setting is going to be removed from the UI. @chaitanyapraveen to confirm, as per our discussion. I've updated the image to remove this setting.

>   <li>For the <strong>Service Account Key Authentication</strong> option, you need to first provide access to the required spreadsheet. <strong>Only the sheets for which you provide the access will be listed in the dropdown</strong>. For more information, refer to the <Link to="#faq">FAQ</Link> section below.</li>

Reviewer 2: @ameypv-rudder instead of sheet, it should be spreadsheets in the below line. 
`Only the sheets for which you provide the access `

We need to be very clear in what is referenced as sheet vs spreadsheet, as this could end up confusing the users.

Author: Absolutely, @nidhilashkari17 - I have made the change.

>   <li>For the <strong>Service Account Key Authentication</strong> option, you need to first provide access to the required spreadsheet. <strong>Only the sheets for which you provide the access will be listed in the dropdown</strong>. For more information, refer to the <Link to="#faq">FAQ</Link> section below.</li>

Author: ```suggestion
  <li>For the <strong>Service Account Key Authentication</strong> option, you need to first provide access to the required spreadsheet. <strong>Only the spreadsheets for which you provide the access will be listed in the dropdown</strong>. For more information, refer to the <Link to="#faq">FAQ</Link> section below.</li>
```

</div>

Google Sheets is now configured as a source. RudderStack will start ingesting data from Google Sheets as per your specified schedule and frequency.

You can further connect this source to your data warehouse by clicking the **Add Destination** button, as shown:

<img src="../../assets/cloud-extract-sources/add-destination.png" alt="Adding a destination" />

<div class="infoBlock">
Use the <strong>Use Existing Destination</strong> option if you have an already-configured data warehouse destination in RudderStack. To configure a data warehouse destination from scratch, select the <strong>Create New Destination</strong> button.
</div>

## How RudderStack syncs data to the warehouse

If you select multiple sheets in the **Select data to import** window, RudderStack creates multiple tables (corresponding to each sheet) in the warehouse.

<img src="../../assets/cloud-extract-sources/googlesheets-beta-sheet.png" alt="Selecting the data to import" />

In the above example, RudderStack creates a table corresponding to each sheet (`List of Tracking Plans`, `Import Template`, and so on) in the warehouse.

> While syncing data to the warehouse, RudderStack creates columns for only those headers (first rows in the sheet) which also have data present in the second row and beyond.

Reviewer: ```suggestion
While syncing data to the warehouse, RudderStack creates columns only for the headers (first rows in the sheet) which also have data present in the second row (mandatory) and beyond (optional).
```

<div class="warningBlock">
The mere presence of a header <strong>does not</strong> create a column in the destination.
</div>

## FAQ

### How do I provide service account access to a Google sheet?

> If you're using a <a href="https://cloud.google.com/iam/docs/creating-managing-service-account-keys#creating">service account</a> to authenticate RudderStack to sync your Google Sheets data, you must also give it the required access to the required spreadsheet. Follow these steps to provide the required sheet access:

Reviewer: ```suggestion
If you're using a <a href="https://cloud.google.com/iam/docs/creating-managing-service-account-keys#creating">service account</a> to authenticate RudderStack to sync your Google Sheets data, you must also give the required access to the service account to access the required spreadsheet. Follow these steps to provide access to the required spreadsheet:
```
